File: NetDeck/IphoneStartViewController.m
//
//  IphoneStartViewController.m
//  Net Deck
//
//  Created by Gereon Steffens on 09.08.15.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "IphoneStartViewController.h"
#import "EditDeckViewController.h"
#import "IphoneIdentityViewController.h"
#import "SettingsViewController.h"
#import "ImportDecksViewController.h"

@interface IphoneStartViewController ()

@property NSMutableArray* runnerDecks;
@property NSMutableArray* corpDecks;

@property NSArray* decks;
@property SettingsViewController* settings;

@property UIBarButtonItem* addButton;
@property UIBarButtonItem* importButton;

@property UIBarButtonItem* settingsButton;
@property UIBarButtonItem* sortButton;

@property NRDeckListSort deckListSort;
@property NSString* filterText;

@end

@implementation IphoneStartViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.delegate = self;
    self.title = @"Net Deck";
    self.tableViewController.title = @"Net Deck";
    
    NSNotificationCenter* nc = [NSNotificationCenter defaultCenter];
    [nc addObserver:self selector:@selector(loadCards:) name:Notifications.loadCards object:nil];
    [nc addObserver:self selector:@selector(importDeckFromClipboard:) name:Notifications.importDeck object:nil];
    
    self.view.backgroundColor = [UIColor colorWithPatternImage:[ImageCache hexTile]];
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.tableView.backgroundColor = [UIColor clearColor];
    
    self.tableView.emptyDataSetDelegate = self;
    self.tableView.emptyDataSetSource = self;
    
    self.searchBar.delegate = self;
    
    BOOL cardsAvailable = [CardManager cardsAvailable];
    self.addButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(createNewDeck:)];
    self.addButton.enabled = cardsAvailable;
    self.importButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"702-import"] style:UIBarButtonItemStylePlain target:self action:@selector(importDecks:)];
    self.importButton.enabled = cardsAvailable;

    UINavigationItem* topItem = self.navigationBar.topItem;
    topItem.rightBarButtonItems = @[ self.addButton, self.importButton ];
    
    self.settingsButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"740-gear"] style:UIBarButtonItemStylePlain target:self action:@selector(openSettings:)];
    self.sortButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"890-sort-ascending-toolbar"] style:UIBarButtonItemStylePlain target:self action:@selector(changeSort:)];
    self.sortButton.enabled = cardsAvailable;
    
    topItem.rightBarButtonItems = @[ self.addButton, self.importButton ];
    topItem.leftBarButtonItems = @[ self.settingsButton, self.sortButton ];
    
    self.deckListSort = [[NSUserDefaults standardUserDefaults] integerForKey:SettingsKeys.DECK_FILTER_SORT];
    
    if ([CardManager cardsAvailable] && [PackManager packsAvailable])
    {
        [self initializeDecks];
    }
    
    self.tableView.contentInset = UIEdgeInsetsZero; // wtf is this needed since iOS9?
    
    [self.tableView setContentOffset:CGPointMake(0, self.searchBar.frame.size.height) animated:NO];
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [CardUpdateCheck checkCardUpdateAvailable:self];
}

-(void) dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

// this is my poor man's replacement for viewWillAppear - I can't figure out why this isn't called when this view is
// back on top :(
-(void) navigationController:(UINavigationController *)navigationController willShowViewController:(UIViewController *)viewController animated:(BOOL)animated
{
    if (viewController != self.tableViewController)
    {
        return;
    }
    
    NSAssert(navigationController.viewControllers.count == 1, @"nav oops");
    if ([CardManager cardsAvailable] && [PackManager packsAvailable])
    {
        [self initializeDecks];
    }
    
    [self.tableView reloadData];
}

-(void) initializeDecks
{
    self.runnerDecks = [DeckManager decksForRole:NRRoleRunner].mutableCopy;
    self.corpDecks = [DeckManager decksForRole:NRRoleCorp].mutableCopy;

    if (self.deckListSort != NRDeckListSortByDate)
    {
        self.runnerDecks = [self sortDecks:self.runnerDecks];
        self.corpDecks = [self sortDecks:self.corpDecks];
        self.decks = @[ self.runnerDecks, self.corpDecks ];
    }
    else
    {
        NSMutableArray* arr = [NSMutableArray arrayWithArray:self.runnerDecks];
        [arr addObjectsFromArray:self.corpDecks];
        self.runnerDecks = [self sortDecks:arr];
        self.corpDecks = [NSMutableArray array];
        self.decks = @[ self.runnerDecks ];
    }
    
    if (self.filterText.length > 0)
    {
        NSPredicate* namePredicate = [NSPredicate predicateWithFormat:@"name CONTAINS[cd] %@", self.filterText];
        
        [self.runnerDecks filterUsingPredicate:namePredicate];
        [self.corpDecks filterUsingPredicate:namePredicate];
    }
    
    BOOL cardsAvailable = [CardManager cardsAvailable];
    self.addButton.enabled = cardsAvailable;
    self.importButton.enabled = cardsAvailable;
    self.sortButton.enabled = cardsAvailable;
    
    NSMutableArray* allDecks = [NSMutableArray arrayWithArray:self.runnerDecks];
    [allDecks addObjectsFromArray:self.corpDecks];
    [[NRDB sharedInstance] updateDeckMap:allDecks];
}

-(void) loadCards:(id)notification
{
    [self initializeDecks];
    [self.tableView reloadData];
}

#pragma mark - import deck

-(void)importDeckFromClipboard:(NSNotification*) notification
{
    NSDictionary* userInfo = notification.userInfo;
    Deck* deck = [userInfo objectForKey:@"deck"];
    
    [deck saveToDisk];
    
    EditDeckViewController* edit = [[EditDeckViewController alloc] initWithNibName:@"EditDeckViewController" bundle:nil];
    edit.deck = deck;
    
    if (self.viewControllers.count > 1)
    {
        [self popToRootViewControllerAnimated:NO];
    }
    [self pushViewController:edit animated:YES];
}

#pragma mark - add new deck

-(void) createNewDeck:(id)sender
{
    UIAlertController* alert = [UIAlertController actionSheetWithTitle:l10n(@"New Deck") message:nil];
    
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"New Runner Deck") handler:^(UIAlertAction *action) {
        [self addNewDeck:NRRoleRunner];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"New Corp Deck") handler:^(UIAlertAction *action) {
        [self addNewDeck:NRRoleCorp];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"Cancel") handler:nil]];
    
    [self presentViewController:alert animated:YES completion:nil];
}

-(void) addNewDeck:(NSInteger)role // actually NRRole
{
    IphoneIdentityViewController* idvc = [[IphoneIdentityViewController alloc] initWithNibName:@"IphoneIdentityViewController" bundle:nil];
    idvc.role = (NRRole)role;
    [self pushViewController:idvc animated:YES];
}

#pragma mark - import

-(void) importDecks:(id)sender
{
    NSUserDefaults* settings = [NSUserDefaults standardUserDefaults];
    BOOL useNrdb = [settings boolForKey:SettingsKeys.USE_NRDB];
    BOOL useDropbox = [settings boolForKey:SettingsKeys.USE_DROPBOX];
    
    if (useNrdb && useDropbox)
    {
        UIAlertController* alert = [UIAlertController alertWithTitle:l10n(@"Import Decks") message:nil];
        
        [alert addAction:[UIAlertAction actionWithTitle:l10n(@"From Dropbox") handler:^(UIAlertAction *action) {
            [self importDecksFrom:NRImportSourceDropbox];
        }]];
        [alert addAction:[UIAlertAction actionWithTitle:l10n(@"From NetrunnerDB.com") handler:^(UIAlertAction *action) {
            [self importDecksFrom:NRImportSourceNetrunnerDb];
        }]];
        [alert addAction:[UIAlertAction cancelAction:nil]];
        
        [self presentViewController:alert animated:NO completion:nil];
    }
    else if (useNrdb)
    {
        [self importDecksFrom:NRImportSourceNetrunnerDb];
    }
    else if (useDropbox)
    {
        [self importDecksFrom:NRImportSourceDropbox];
    }
    else
    {
        UIAlertController* alert = [UIAlertController alertWithTitle:l10n(@"Import Decks")
                                                             message:l10n(@"Connect to your Dropbox and/or NetrunnerDB.com account first.")];
        
        [alert addAction:[UIAlertAction actionWithTitle:l10n(@"OK") handler:nil]];
        
        [self presentViewController:alert animated:NO completion:nil];
    }
}

-(void) importDecksFrom:(NRImportSource)importSource
{
    ImportDecksViewController* import = [[ImportDecksViewController alloc] init];
    import.source = importSource;
    [self pushViewController:import animated:YES];
}

#pragma mark - sort

-(void) changeSort:(id)sender
{
    UIAlertController* alert = [UIAlertController actionSheetWithTitle:l10n(@"Sort by") message:nil];
    
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"Date") handler:^(UIAlertAction *action) {
        [self changeSortType:NRDeckListSortByDate];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"Faction") handler:^(UIAlertAction *action) {
        [self changeSortType:NRDeckListSortByFaction];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"A-Z") handler:^(UIAlertAction *action) {
        [self changeSortType:NRDeckListSortByName];
    }]];
    [alert addAction:[UIAlertAction cancelAction:nil]];
    
    [self presentViewController:alert animated:NO completion:nil];
}

-(void) changeSortType:(NRDeckListSort)sort
{
    [[NSUserDefaults standardUserDefaults] setInteger:sort forKey:SettingsKeys.DECK_FILTER_SORT];
    self.deckListSort = sort;
    
    [self initializeDecks];
    [self.tableView reloadData];
}

-(NSMutableArray<Deck*>*) sortDecks:(NSArray<Deck*>*)decks
{
    switch (self.deckListSort)
    {
        case NRDeckListSortByName:
            decks = [decks sortedArrayUsingComparator:^NSComparisonResult(Deck* d1, Deck* d2) {
                return [[d1.name lowercaseString] compare:[d2.name lowercaseString]];
            }];
            break;
        case NRDeckListSortByDate:
            decks = [decks sortedArrayUsingComparator:^NSComparisonResult(Deck* d1, Deck* d2) {
                NSComparisonResult cmp = [d2.lastModified compare:d1.lastModified];
                if (cmp == NSOrderedSame)
                {
                    return [[d1.name lowercaseString] compare:[d2.name lowercaseString]];
                }
                return cmp;
            }];
            break;
        case NRDeckListSortByFaction:
            decks = [decks sortedArrayUsingComparator:^NSComparisonResult(Deck* d1, Deck* d2) {
                NSString* faction1 = [Faction name:d1.identity.faction];
                NSString* faction2 = [Faction name:d2.identity.faction];
                NSComparisonResult cmp = [faction1 compare:faction2];
                if (cmp == NSOrderedSame)
                {
                    cmp = [[d1.identity.name lowercaseString] compare:[d2.identity.name lowercaseString]];
                    if (cmp == NSOrderedSame)
                    {
                        return [[d1.name lowercaseString] compare:[d2.name lowercaseString]];
                    }
                    return cmp;
                }
                return cmp;
            }];
            break;
    }
    
    return [decks mutableCopy];
}

#pragma mark - settings

-(void) openSettings:(id)sender
{
    self.settings = [[SettingsViewController alloc] init];
    [self pushViewController:self.settings.iask animated:YES];
}

#pragma mark - browser

-(void) titleButtonTapped:(id)sender
{
    BrowserViewController* browser = [[BrowserViewController alloc] initWithNibName:@"BrowserViewController" bundle:nil];
    [self pushViewController:browser animated:YES];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.decks.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSArray* arr = self.decks[section];
    return arr.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString* cellIdentifier = @"deckCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    
    NSArray* decks = self.decks[indexPath.section];
    Deck* deck = decks[indexPath.row];
    
    cell.textLabel.text = deck.name;
    
    if (deck.identity)
    {
        cell.detailTextLabel.text = deck.identity.name;
        cell.detailTextLabel.textColor = deck.identity.factionColor;
    }
    else
    {
        cell.detailTextLabel.text = l10n(@"No Identity");
        cell.detailTextLabel.textColor = [UIColor blackColor];
    }
    
    return cell;
}

-(void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    Deck* deck = [self.decks objectAtIndexPath:indexPath];
    
    EditDeckViewController* edit = [[EditDeckViewController alloc] initWithNibName:@"EditDeckViewController" bundle:nil];
    edit.deck = deck;
    
    [self pushViewController:edit animated:YES];
}

-(NSString*) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if (self.decks.count == 1)
    {
        return nil;
    }
    return section == 0 ? l10n(@"Runner") : l10n(@"Corp");
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete)
    {
        NSMutableArray* decks = self.decks[indexPath.section];
        Deck* deck = decks[indexPath.row];
        
        [decks removeObjectAtIndex:indexPath.row];
        [[NRDB sharedInstance] deleteDeck:deck.netrunnerDbId];
        [DeckManager removeFile:deck.filename];
        
        [self.tableView beginUpdates];
        [self.tableView deleteRowsAtIndexPaths:@[ indexPath ] withRowAnimation:UITableViewRowAnimationLeft];
        [self.tableView endUpdates];
    }
}

#pragma mark search bar

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    self.filterText = searchText;
    [self initializeDecks];
    [self.tableView reloadData];
}

-(void) searchBarTextDidEndEditing:(UISearchBar *)searchBar
{
    searchBar.showsCancelButton = NO;
    self.tableView.tableHeaderView = self.searchBar;
}

-(void) searchBarTextDidBeginEditing:(UISearchBar *)searchBar
{
    searchBar.showsCancelButton = YES;
    self.tableView.tableHeaderView = self.searchBar;
}

-(void) searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    [searchBar resignFirstResponder];
}

#pragma mark 3d touch shortcuts

-(void) openBrowser {
    [self titleButtonTapped:nil];
}

#pragma mark empty dataset

-(NSAttributedString*) titleForEmptyDataSet:(UIScrollView *)scrollView
{
    NSDictionary *attributes = @{NSFontAttributeName: [UIFont systemFontOfSize:21.0f],
                                 NSForegroundColorAttributeName: [UIColor lightGrayColor]};
    
    BOOL cardsAvailable = [CardManager cardsAvailable] && [PackManager packsAvailable];
    NSString* title = cardsAvailable ? @"No Decks" : @"No Card Data";
    
    return [[NSAttributedString alloc] initWithString:l10n(title) attributes:attributes];
}

- (NSAttributedString *)descriptionForEmptyDataSet:(UIScrollView *)scrollView
{
    BOOL cardsAvailable = [CardManager cardsAvailable] && [PackManager packsAvailable];
    NSString *text = cardsAvailable ? @"Your decks will be shown here" : @"To use this app, you must first download card data.";
    
    NSDictionary *attributes = @{NSFontAttributeName: [UIFont systemFontOfSize:14.0f],
                                 NSForegroundColorAttributeName: [UIColor lightGrayColor],
                                 };
    
    return [[NSAttributedString alloc] initWithString:l10n(text) attributes:attributes];
}

- (NSAttributedString*) buttonTitleForEmptyDataSet:(UIScrollView *)scrollView forState:(UIControlState)state {
    UIColor* color = self.tableView.tintColor;
    NSDictionary *attributes = @{NSFontAttributeName: [UIFont systemFontOfSize:17.0f], NSForegroundColorAttributeName: color };
    BOOL cardsAvailable = [CardManager cardsAvailable] && [PackManager packsAvailable];
    NSString* text = cardsAvailable ? @"New Deck" : @"Download";
    return [[NSAttributedString alloc] initWithString:l10n(text) attributes:attributes];
}

-(UIColor*) backgroundColorForEmptyDataSet:(UIScrollView *)scrollView {
    return [UIColor colorWithPatternImage:[ImageCache hexTileLight]];
}

-(void) emptyDataSet:(UIScrollView *)scrollView didTapButton:(UIButton *)button {
    BOOL cardsAvailable = [CardManager cardsAvailable] && [PackManager packsAvailable];
    if (cardsAvailable) {
        [self createNewDeck:nil];
    } else {
        [DataDownload downloadCardData];
    }
}

@end

File: NetDeck/IphoneStartViewController.swift
//
//  IphoneStartViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 31.10.16.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

import UIKit
import DZNEmptyDataSet
import SwiftyUserDefaults
import EasyTipView

class IphoneStartViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, UISearchBarDelegate, DZNEmptyDataSetDelegate, DZNEmptyDataSetSource, StartViewController {

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var searchBar: UISearchBar!
    
    private var decks = [[Deck]]()
    
    private var addButton: UIBarButtonItem!
    private var importButton: UIBarButtonItem!
    fileprivate var settingsButton: UIBarButtonItem!
    private var sortButton: UIBarButtonItem!
    fileprivate var titleButton: UIButton!
    
    private var deckListSort = DeckListSort.byName
    private var filterText = ""
    
    private var keyboardObserver: KeyboardObserver!
    fileprivate var tipView: EasyTipView?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "Net Deck"
        
        let nc = NotificationCenter.default
        nc.addObserver(self, selector: #selector(self.loadCards(_:)), name: Notifications.loadCards, object: nil)
        nc.addObserver(self, selector: #selector(self.importDeckFromClipboard(_:)), name: Notifications.importDeck, object: nil)
        
        self.view.backgroundColor = UIColor(patternImage: ImageCache.hexTile)
        self.tableView.tableFooterView = UIView(frame: CGRect.zero)
        self.tableView.backgroundColor = .clear
        
        self.tableView.delegate = self
        self.tableView.dataSource = self
        self.tableView.emptyDataSetSource = self
        self.tableView.emptyDataSetDelegate = self
        
        self.searchBar.delegate = self
        
        let cardsAvailable = CardManager.cardsAvailable
        self.addButton = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(self.createNewDeck(_:)))
        self.addButton.isEnabled = cardsAvailable
        
        self.importButton = UIBarButtonItem(image: UIImage(named: "702-import"), style: .plain, target: self, action: #selector(self.importDecks(_:)))
        self.importButton.isEnabled = cardsAvailable
        
        self.settingsButton = UIBarButtonItem(image: UIImage(named: "740-gear"), style: .plain, target: self, action: #selector(self.openSettings(_:)))
        
        self.sortButton = UIBarButtonItem(image: UIImage(named: "890-sort-ascending-toolbar"), style: .plain, target: self, action: #selector(self.changeSort(_:)))
        
        self.titleButton = UIButton(type: .system)
        self.titleButton.frame = CGRect(x: 0, y: 0, width: 0, height: 33)
        self.titleButton.setTitle("Net Deck", for: .normal)
        self.titleButton.titleLabel?.font = UIFont.systemFont(ofSize: 17, weight: UIFont.Weight.medium)
        self.titleButton.addTarget(self, action: #selector(self.openBrowser), for: .touchUpInside)
        
        self.deckListSort = Defaults[.deckFilterSort]
        
        self.tableView.contentInset = UIEdgeInsets.zero
        self.tableView.scrollFix()
        
        self.keyboardObserver = KeyboardObserver(handler: self)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        self.navigationItem.leftBarButtonItems = [ self.settingsButton, self.sortButton ]
        self.navigationItem.rightBarButtonItems = [ self.addButton, self.importButton ]
        
        self.navigationItem.titleView = self.titleButton
        
        if CardManager.cardsAvailable {
            self.initializeDecks()
        }
        
        self.settingsButton.isEnabled = true
        
        self.tableView.reloadData()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        let _ = CardUpdateCheck.checkCardUpdateAvailable()
        
        self.titleButton.sizeToFit()
        self.showTipView()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        self.dismissTipView()
    }
    
    func initializeDecks() {
        var runnerDecks = DeckManager.decksForRole(.runner)
        var corpDecks = DeckManager.decksForRole(.corp)
        
        if self.filterText.count > 0 {
            let namePredicate = NSPredicate(format: "name CONTAINS[cd] %@", self.filterText)
            let identityPredicate = NSPredicate(format: "(identity.name CONTAINS[cd] %@) or (identity.englishName CONTAINS[cd] %@)", self.filterText, self.filterText)
            let predicate = NSCompoundPredicate(orPredicateWithSubpredicates: [namePredicate, identityPredicate])
            runnerDecks = runnerDecks.filter { predicate.evaluate(with: $0) }
            corpDecks = corpDecks.filter { predicate.evaluate(with: $0) }
        }
        
        if self.deckListSort != .byDate {
            runnerDecks = self.sortDecks(runnerDecks)
            corpDecks = self.sortDecks(corpDecks)
            self.decks = [ runnerDecks, corpDecks ]
        } else {
            let decks = runnerDecks + corpDecks
            runnerDecks = self.sortDecks(decks)
            self.decks = [ runnerDecks ]
        }
        
        let cardsAvailable = CardManager.cardsAvailable
        self.addButton.isEnabled = cardsAvailable
        self.importButton.isEnabled = cardsAvailable
        self.sortButton.isEnabled = cardsAvailable
        
        let allDecks = runnerDecks + corpDecks
        NRDB.sharedInstance.updateDeckMap(allDecks)
    }
    
    @objc func loadCards(_ notification: Notification) {
        self.initializeDecks()
        self.tableView.reloadData()
    }
    
    // MARK: - import deck

    @objc func importDeckFromClipboard(_ notification: Notification) {
        guard let deck = notification.userInfo?["deck"] as? Deck else {
            return
        }
        
        deck.saveToDisk()
        
        let edit = EditDeckViewController()
        edit.deck = deck
        
        if let nav = self.navigationController {
            if nav.viewControllers.count > 1 {
                nav.popToRootViewController(animated: false)
            }
            nav.pushViewController(edit, animated: true)
        }
    }
    
    // MARK: - add new deck
    
    @objc func createNewDeck(_ sender: Any) {
        self.createNewDeck()
    }
    
    func createNewDeck() {
        let alert = UIAlertController.actionSheet(title: "New Deck".localized(), message: nil)
        
        alert.addAction(UIAlertAction(title: "New Runner Deck".localized()) { action in
            self.addNewDeck(.runner)
        })
        alert.addAction(UIAlertAction(title: "New Corp Deck".localized()) { action in
            self.addNewDeck(.corp)
        })
        alert.addAction(UIAlertAction(title: "Cancel".localized(), style: .cancel, handler: nil))
        
        self.present(alert, animated: true, completion: nil)
    }
    
    func addNewDeck(_ role: Role) {
        let idvc = IphoneIdentityViewController()
        idvc.role = role
        self.navigationController?.pushViewController(idvc, animated: true)
    }
    
    // MARK: - import
    
    @objc func importDecks(_ sender: UIButton) {
        let useNrdb = Defaults[.useNrdb]
        let useDropbox = Defaults[.useDropbox]
        
        if useNrdb && useDropbox {
            let alert = UIAlertController.actionSheet(title: "Import Decks".localized(), message:nil)
            
            alert.addAction(UIAlertAction(title: "From Dropbox".localized()) { action in
                self.importDecksFrom(.dropbox)
            })
            alert.addAction(UIAlertAction(title: "From NetrunnerDB.com".localized()) { action in
                self.importDecksFrom(.netrunnerDb)
            })
            
            alert.addAction(UIAlertAction(title: "Cancel".localized(), style: .cancel, handler: nil))
            
            self.present(alert, animated: true, completion: nil)
        } else if useNrdb {
            self.importDecksFrom(.netrunnerDb)
        } else if useDropbox {
            self.importDecksFrom(.dropbox)
        } else {
            let alert = UIAlertController.alert(title: "Import Decks".localized(),
                message: "Connect to your Dropbox and/or NetrunnerDB.com account first.".localized())
            
            alert.addAction(UIAlertAction(title: "OK".localized(), handler: nil))
            
            self.present(alert, animated: true, completion:nil)
        }
    }
    
    func importDecksFrom(_ importSource: ImportSource) {
        let importVc = ImportDecksViewController()
        importVc.source = importSource
        self.navigationController?.pushViewController(importVc, animated: true)
    }
    
    // MARK: - sort
    
    @objc func changeSort(_ sender: UIBarButtonItem) {
        let alert = UIAlertController.actionSheet(title: "Sort by".localized(), message:nil)
        
        alert.addAction(UIAlertAction(title: "Date".localized().checked(self.deckListSort == .byDate)) { action in
            self.changeSortType(.byDate)
        })
        alert.addAction(UIAlertAction(title: "Faction".localized().checked(self.deckListSort == .byFaction)) { action in
            self.changeSortType(.byFaction)
        })
        alert.addAction(UIAlertAction(title: "A-Z".localized().checked(self.deckListSort == .byName)) { action in
            self.changeSortType(.byName)
        })
        
        alert.addAction(UIAlertAction(title: "Cancel".localized(), style: .cancel, handler: nil))
        
        self.present(alert, animated: true, completion: nil)
    }
    
    func changeSortType(_ sort: DeckListSort) {
        Defaults[.deckFilterSort] = sort
        self.deckListSort = sort
        
        self.initializeDecks()
        self.tableView.reloadData()
    }
    
    func sortDecks(_ decks: [Deck]) -> [Deck] {
        let result: [Deck]
        switch self.deckListSort {
        case .byName:
            result = decks.sorted { $1.name.lowercased() > $0.name.lowercased() }
        case .byDate:
            result = decks.sorted {
                let d1 = $0.lastModified ?? Date()
                let d2 = $1.lastModified ?? Date()
                if d1 == d2 {
                    return $1.name.lowercased() > $0.name.lowercased()
                } else {
                    return d1 > d2
                }
            }
        case .byFaction:
            result = decks.sorted {
                let faction1 = Faction.name(for: $0.identity?.faction ?? .none) 
                let faction2 = Faction.name(for: $1.identity?.faction ?? .none)
                if faction1 == faction2 {
                    return $1.name.lowercased() > $0.name.lowercased()
                } else {
                    return faction1 > faction2
                }
            }
        }
        
        return result
    }
    
    // MARK: - settings
    
    @objc func openSettings(_ sender: UIBarButtonItem) {
        sender.isEnabled = false
        let settings = Settings.viewController
        if settings != self.navigationController?.topViewController {
            self.navigationController?.pushViewController(settings, animated: true)
        }
    }
    
    // MARK: - browser
    
    func titleButtonTapped(_ sender: UIBarButtonItem) {
        self.openBrowser()
    }
    
    @objc func openBrowser() {
        if CardManager.cardsAvailable {
            let browser = BrowserViewController()
            self.navigationController?.pushViewController(browser, animated: true)
        } else {
            DataDownload.downloadCardData()
        }
    }
    
    // MARK: - table view
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.decks.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.decks[section].count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "deckCell"
        
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier) ?? {
            let c = UITableViewCell(style: .subtitle, reuseIdentifier: cellIdentifier)
            c.selectionStyle = .none
            c.accessoryType = .disclosureIndicator
            return c
        }()
        
        let deck = self.decks[indexPath.section][indexPath.row]
        cell.textLabel?.text = deck.name
        
        if let identity = deck.identity {
            cell.detailTextLabel?.text = identity.name
            cell.detailTextLabel?.textColor = identity.factionColor
        } else {
            cell.detailTextLabel?.text = "No Identity".localized()
            cell.detailTextLabel?.textColor = .black
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let edit = EditDeckViewController()
        edit.deck = self.decks[indexPath.section][indexPath.row]
        
        self.navigationController?.pushViewController(edit, animated: true)
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if self.decks.count == 1 {
            return nil
        }
        return section == 0 ? "Runner".localized() : "Corp".localized()
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            let deck = self.decks[indexPath.section][indexPath.row]
            NRDB.sharedInstance.deleteDeck(deck.netrunnerDbId)
            if let filename = deck.filename {
                DeckManager.removeFile(filename)
            }
            
            self.decks[indexPath.section].remove(at: indexPath.row)
            
            self.tableView.beginUpdates()
            self.tableView.deleteRows(at: [ indexPath ], with: .left)
            self.tableView.endUpdates()
        }
    }
    
    // MARK: - search bar
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        self.filterText = searchText
        self.initializeDecks()
        self.tableView.reloadData()
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        self.searchBar.setShowsCancelButton(false, animated: true)
    }
    
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        self.searchBar.setShowsCancelButton(true, animated: true)
    }
    
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        self.searchBar.resignFirstResponder()
    }

    // MARK: - empty dataset
    
    func title(forEmptyDataSet scrollView: UIScrollView!) -> NSAttributedString! {
        if self.filterText.count > 0 {
            return nil
        }
        
        let attrs: [NSAttributedStringKey: Any] = [ NSAttributedStringKey.font: UIFont.systemFont(ofSize: 21.0), NSAttributedStringKey.foregroundColor: UIColor.lightGray ]
        
        let title = CardManager.cardsAvailable ? "No Decks" : "No Card Data"
        
        return NSAttributedString(string: title.localized(), attributes: attrs)
    }
    
    func description(forEmptyDataSet scrollView: UIScrollView!) -> NSAttributedString! {
        if self.filterText.count > 0 {
            return nil
        }

        let text = CardManager.cardsAvailable ? "Your decks will be shown here" : "To use this app, you must first download card data."

        let attrs: [NSAttributedStringKey: Any] = [ NSAttributedStringKey.font: UIFont.systemFont(ofSize: 14.0), NSAttributedStringKey.foregroundColor: UIColor.lightGray ]
        
        return NSAttributedString(string: text.localized(), attributes: attrs)
    }
    
    func buttonTitle(forEmptyDataSet scrollView: UIScrollView!, for state: UIControlState) -> NSAttributedString! {
        if self.filterText.count > 0 {
            return nil
        }
        
        let color = self.tableView.tintColor ?? .blue
        let attrs: [NSAttributedStringKey: Any] = [ NSAttributedStringKey.font: UIFont.systemFont(ofSize: 17.0), NSAttributedStringKey.foregroundColor: color ]
        
        let text = CardManager.cardsAvailable ? "New Deck" : "Download"
        
        return NSAttributedString(string: text.localized(), attributes: attrs)
    }
    
    func backgroundColor(forEmptyDataSet scrollView: UIScrollView!) -> UIColor! {
        return UIColor(patternImage: ImageCache.hexTileLight)
    }
    
    func emptyDataSet(_ scrollView: UIScrollView!, didTap button: UIButton!) {
        if CardManager.cardsAvailable {
            self.createNewDeck()
        } else {
            DataDownload.downloadCardData()
        }
    }
    
}

// MARK: - keyboard
extension IphoneStartViewController: KeyboardHandling {
    func keyboardWillShow(_ info: KeyboardInfo) {
        let screenHeight = UIScreen.main.bounds.size.height
        let kbHeight = screenHeight - info.endFrame.origin.y
        
        let insets = UIEdgeInsets(top: 0, left: 0, bottom: kbHeight, right: 0)
        
        self.tableView.contentInset = insets
        self.tableView.scrollIndicatorInsets = insets
    }
    
    func keyboardWillHide(_ info: KeyboardInfo) {
        let insets = UIEdgeInsets.zero
        self.tableView.contentInset = insets
        self.tableView.scrollIndicatorInsets = insets
    }
}

// MARK: - tip view
extension IphoneStartViewController {
    
    fileprivate func showTipView() {
        if CardManager.cardsAvailable && !Defaults[.browserHintShown] {
            self.presentTipView()
            Defaults[.browserHintShown] = true
        }
    }
    
    private func presentTipView() {
        var prefs = EasyTipView.Preferences()
        prefs.drawing.font = UIFont.systemFont(ofSize: 13, weight: UIFont.Weight.thin)
        prefs.drawing.cornerRadius = 5
        prefs.drawing.foregroundColor = .white
        prefs.drawing.backgroundColor = .darkGray
        prefs.drawing.arrowPosition = .top
        
        let browserTip = "Net Deck also offers a card browser.\nTap the title to open it.".localized()
        
        self.tipView = EasyTipView(text: browserTip, preferences: prefs)
        
        let view = self.titleButton
        
        self.tipView?.show(animated: true, forView: view!)
    }
    
    fileprivate func dismissTipView() {
        self.tipView?.dismiss()
    }
}
